Unified inputs for grayscale ops and transforms

vfdev-5 opened the pull request:
Related to #2292

Description:
- Unified inputs for grayscale op and transforms: `GrayScale`, `RandomGrayscale`, `F.rgb_to_grayscale`.
- Modified `F_t.rgb_to_grayscale` implemention to match better PIL implementation
  - Side effect: this breaks `test_adjustments` => we need to update this test according to how we test results in `test_adjust_gamma`.
- Deprecated `F.to_grayscale` in favor of `F.rgb_to_grayscale`


EDIT: 
- uint8 type is transformed as `L = (19595 * r + 38470 * g + 7471 * b + 2 ** 15) / 2 ** 16`
- How to transform floating point ? -> `(0.299 * r + 0.587 * g + 0.114 * b)`

Unlocks #2566 #2595

vfdev-5 commented:
@pytorchbot retest this please

fmassa commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -67,22 +76,51 @@ def crop(img: Tensor, top: int, left: int, height: int, width: int) -> Tensor:
+    if num_output_channels not in (1, 3):
+        raise ValueError('num_output_channels should be either 1 or 3')
+
+    r = img[..., 0, :, :].float()
+    g = img[..., 1, :, :].float()
+    b = img[..., 2, :, :].float()
+    if not img.is_floating_point():
+        # According to PIL docs: PIL grayscale L mode is L = R * 299/1000 + G * 587/1000 + B * 114/1000
+        # but implementation is slightly different:
+        # https://github.com/python-pillow/Pillow/blob/4634eafe3c695a014267eefdce830b4a825beed7/
+        # src/libImaging/Convert.c#L47
+        # ((rgb)[0]*19595 + (rgb)[1]*38470 + (rgb)[2]*7471 + 0x8000) >> 16
+        # l_img = ((19595 * r + 38470 * g + 7471 * b + 2 ** 15) / 2 ** 16).to(img.dtype)
+        l_img = torch.floor((19595 * r + 38470 * g + 7471 * b + 2 ** 15) / 2 ** 16).to(img.dtype)
+    else:
+        l_img = (0.299 * r + 0.587 * g + 0.114 * b).to(img.dtype)
```
I'm not sure we should be following PIL (approximate) implementation closely here.

The numbers they are using are just integer approximations of the official conversion numbers (so that they can avoid casting to float and back to `uint8`).

Our implementation closely follows [the TF one](https://github.com/tensorflow/tensorflow/blob/v2.3.0/tensorflow/python/ops/image_ops_impl.py#L2105-L2138).

One thing worth discussion though is first factor being 0.2989 vs 0.299 (both OpenCV and PIL uses 0.299), which we might want to consider changing potentially (but @pedrofreire might know better where he took the 0.2989 from)

vfdev-5 replied:
> One thing worth discussion though is first factor being 0.2989 vs 0.299 (both OpenCV and PIL uses 0.299), which we might want to consider changing potentially (but @pedrofreire might know better where he took the 0.2989 from)

Yes, this is a bit strange as tf code says that they follow a wikipedia page (https://en.wikipedia.org/wiki/Luma_%28video%29) and where now it is also 0.299 .

> I'm not sure we should be following PIL (approximate) implementation closely here.

You would like to have a single way to compute gray image (`0.299 * r + 0.587 * g + 0.114 * b`) and adapt the tests ?

fmassa commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -67,22 +76,51 @@ def crop(img: Tensor, top: int, left: int, height: int, width: int) -> Tensor:
+    else:
+        l_img = (0.299 * r + 0.587 * g + 0.114 * b).to(img.dtype)
+
+    if num_output_channels == 3:
+        l_img = torch.stack([l_img, l_img, l_img], dim=-3)
+    else:
+        l_img = l_img.unsqueeze(dim=-3)
```
Do we want the `num_output_channels` argument to this function? I'm not sure it adds much value.

And if we do, shouldn't we instead use a more optimized implementation, like
```python
l_img = l_img.unsqueeze(dim=-3).expand(...)
```

vfdev-5 replied:
`num_output_channels` is originally `F_pil.to_grayscale` and now in `F_pil.rgb_to_grayscale`. So, I'd say to keep compatility we might keep it. Yes, if `l_img = l_img.unsqueeze(dim=-3).expand(...)` works faster doing the same thing I agree that we have to use this implementation.

fmassa commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -67,22 +76,51 @@ def crop(img: Tensor, top: int, left: int, height: int, width: int) -> Tensor:
 
     Args:
         img (Tensor): Image to be converted to Grayscale in the form [C, H, W].
+        num_output_channels (int): number of channels of the output image. Value can be 1 or 3. Default, 1.
```
I wonder if we need this at all or if we should move this logic to the transform itself?

vfdev-5 replied:
Currently, `Grayscale` transform is using `num_output_channels` as argument and `RandomGrayscale` outputs the same number of channels as in the input.
If you are OK with BC change, we can remove it. It is a bit difficult to estimage how many users apply `F.to_grayscale` with this option.

fmassa commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -67,22 +76,51 @@ def crop(img: Tensor, top: int, left: int, height: int, width: int) -> Tensor:
+        l_img = l_img.unsqueeze(dim=-3)
 
-    return (0.2989 * img[0] + 0.5870 * img[1] + 0.1140 * img[2]).to(img.dtype)
+    return l_img
```
I agree though that we should always return a tensor with the same number of dimensions as the original tensor

fmassa commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -67,22 +76,51 @@ def crop(img: Tensor, top: int, left: int, height: int, width: int) -> Tensor:
+    if num_output_channels not in (1, 3):
+        raise ValueError('num_output_channels should be either 1 or 3')
+
+    r = img[..., 0, :, :].float()
+    g = img[..., 1, :, :].float()
+    b = img[..., 2, :, :].float()
```
if we need the `rgb` values, what about doing
```python
r, g, b = img.unbind(dim=-3)
```

fmassa reviewed (commented):
Thanks for the PR Victor!

I've made a few comments, let me know what you think

vfdev-5 commented:
@fmassa I updated the PR. Changes:
- `F.to_grayscale` works on PIL Image and goes directly to `F_pil.to_grayscale`. 
- `F.rgb_to_grayscale` works on both: PIL images and Tensors and dispatch to `F_pil.to_grayscale` or `F_t.rgb_to_grayscale`
- no more deprecations
- reverted to unique manner to compute `F_t.rgb_to_grayscale` on float/uint8 tensors and added other review comments 
- kept `num_output_channels` in `F.to_grayscale`, `F.rgb_to_grayscale`. We can iterate on that.


Let me know if it corresponds to your ideas.

fmassa commented on torchvision/transforms/functional_tensor.py:
```diff
@@ -67,22 +76,41 @@ def crop(img: Tensor, top: int, left: int, height: int, width: int) -> Tensor:
+    # This implementation closely follows the TF one:
+    # https://github.com/tensorflow/tensorflow/blob/v2.3.0/tensorflow/python/ops/image_ops_impl.py#L2105-L2138
+    l_img = (0.2989 * r + 0.587 * g + 0.114 * b).to(img.dtype)
+    l_img = l_img.unsqueeze(dim=-3)
```
this is a BC-breaking change, as before the tensor returned would have 2 dimensions. I'll change the title of the PR to account for this

fmassa reviewed (approved):
Thanks!
